File: examples/海外呼叫中心/代码示例/Java/AgentController.java
package com.example.callcenter.service;

import org.freeswitch.esl.client.inbound.Client;
import org.freeswitch.esl.client.inbound.InboundConnectionFailure;
import org.freeswitch.esl.client.transport.message.EslMessage;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

/**
 * 坐席呼叫控制器 (基于 ESL)
 * 依赖库: org.freeswitch.esl.client:org.freeswitch.esl.client:0.9.2
 */
@Service
public class AgentController {

    private static final Logger logger = LoggerFactory.getLogger(AgentController.class);
    private final Client eslClient = new Client();

    public AgentController() {
        try {
            // 连接到 FreeSWITCH ESL 端口 (默认 8021)
            eslClient.connect("127.0.0.1", 8021, "ClueCon", 10);
        } catch (InboundConnectionFailure e) {
            logger.error("连接 FreeSWITCH 失败", e);
        }
    }

    /**
     * 发起点击拨号 (Click-to-Dial)
     * 逻辑: 先呼叫坐席(WebRTC)，坐席接通后，再呼叫客户(Aliyun)
     *
     * @param agentId 坐席分机号 (如 1001)
     * @param customerNumber 客户号码 (如 00852...)
     */
    public void clickToDial(String agentId, String customerNumber) {
        // 构造 originate 命令
        // 格式: originate {origination_caller_id_number=...}user/1001 &bridge(sofia/gateway/aliyun/...)
        
        String command = String.format(
            "originate {origination_caller_id_name='System',origination_caller_id_number='%s'}user/%s &bridge(sofia/gateway/aliyun-intl/%s)",
            customerNumber, agentId, customerNumber
        );

        logger.info("执行外呼: {}", command);

        EslMessage response = eslClient.sendSyncApiCommand(command, null);
        
        if (response.getBodyLines().stream().anyMatch(line -> line.contains("+OK"))) {
            logger.info("外呼指令发送成功，UUID: {}", response.getHeaderValue("Job-UUID"));
        } else {
            logger.error("外呼失败: {}", response.getBodyLines());
        }
    }
    
    /**
     * 监听坐席状态 (简易版)
     */
    public void subscribeEvents() {
        // 订阅通道状态和心跳
        eslClient.setEventSubscriptions("plain", "CHANNEL_CREATE CHANNEL_HANGUP HEARTBEAT");
        
        // 在实际应用中，需要通过 addEventListener 注册回调处理逻辑
    }
}

File: examples/海外呼叫中心/代码示例/Rust/agent_controller.rs
use tokio::net::TcpStream;
use tokio::io::{AsyncReadExt, AsyncWriteExt};
use std::error::Error;

/// Rust ESL 客户端示例
/// 演示如何连接 FreeSWITCH 并发送外呼指令
#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    // 1. 连接 FreeSWITCH ESL 端口
    let mut stream = TcpStream::connect("127.0.0.1:8021").await?;
    println!("已连接到 FreeSWITCH");

    // 2. 读取欢迎消息 (Content-Type: auth/request)
    let mut buffer = [0; 1024];
    let n = stream.read(&mut buffer).await?;
    println!("收到: {}", String::from_utf8_lossy(&buffer[..n]));

    // 3. 发送认证指令
    let auth_cmd = "auth ClueCon\n\n";
    stream.write_all(auth_cmd.as_bytes()).await?;

    // 4. 读取认证结果 (+OK accepted)
    let n = stream.read(&mut buffer).await?;
    let response = String::from_utf8_lossy(&buffer[..n]);
    if !response.contains("+OK accepted") {
        panic!("认证失败");
    }
    println!("认证成功");

    // 5. 发送点击拨号指令 (bgapi originate)
    // 场景: 坐席 1001 呼叫香港号码 0085298765432
    let agent_id = "1001";
    let customer_number = "0085298765432";
    
    let originate_cmd = format!(
        "bgapi originate {{origination_caller_id_number={}}}user/{} &bridge(sofia/gateway/aliyun-intl/{})\n\n",
        customer_number, agent_id, customer_number
    );

    println!("发送指令: {}", originate_cmd);
    stream.write_all(originate_cmd.as_bytes()).await?;

    // 6. 读取指令响应 (Job-UUID)
    let n = stream.read(&mut buffer).await?;
    println!("指令响应: {}", String::from_utf8_lossy(&buffer[..n]));

    Ok(())
}
